auth/permission_details.cc

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ *
+ * Modified by ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#include "auth/permission_details.hh"
+
+// FMT.
+#include <fmt/format.h>
+
+// Seastar features.
+#include <seastar/coroutine/maybe_yield.hh>
+
+// Scylla includes.
+#include "auth/permission.hh"
+#include "auth/resource.hh"
+#include "cql3/util.hh"
+#include "log.hh"
+
+// STD.
+#include <ranges>
+#include <string_view>
+
+// Note: The name of a keyspace or a table can only consist of alphanumeric types.
+
+namespace auth {
+
+namespace {
+
+logging::logger pdlogger{"permission_details"};
+
+// Precondition: `permission` and `role` can be safely used in CQL statements without (further)
+//               wrapping them in quotation marks.
+sstring describe_data(std::string_view permission, const resource& r, std::string_view role) {
+    const auto view = data_resource_view(r);
+    const auto maybe_ks = view.keyspace();
+    const auto maybe_cf = view.table();
+
+    // The documentation says:
+    //
+    //     Both keyspace and table names consist of only alphanumeric characters, cannot be empty,
+    //     and are limited in size to 48 characters (that limit exists mostly to avoid filenames,
+    //     which may include the keyspace and table name, to go over the limits of certain file systems).
+    //     By default, keyspace and table names are case insensitive (myTable is equivalent to mytable),
+    //     but case sensitivity can be forced by using double-quotes ("myTable" is different from mytable).
+    //
+    // That's why we wrap identifiers with quotation marks below.
+
+    if (!maybe_ks) {
+        return seastar::format("GRANT {} ON ALL KEYSPACES TO {};", permission, role);
+    }
+    const auto ks = cql3::util::maybe_quote(*maybe_ks);
+
+    if (!maybe_cf) {
+        return seastar::format("GRANT {} ON KEYSPACE {} TO {};", permission, ks, role);
+    }
+    const auto cf = cql3::util::maybe_quote(*maybe_cf);
+
+    return seastar::format("GRANT {} ON {}.{} TO {};", permission, ks, cf, role);
+}
+
+// Precondition: `permission` and `role` can be safely used in CQL statements without (further)
+//               wrapping them in quotation marks.
```

Comment:
Does this imply that the caller has to escape/quote the `permission`/`role` parameter before passing it to the function? It's a bad interface in my opinion. It's much more natural to pass "raw" names around and only escape/quote them before printing them. Given that, I think it is easy to misuse because the precondition is not statically checked by the compiler.